File: UVa 101 The Blocks Problem/main.go
// UVa 101 - The Blocks Problem
package main

import (
	"fmt"
	"os"
)

// Matrix holds all blocks.
type Matrix struct {
	blocks [][]int
}

func newBlocks(n int) [][]int {
	b := make([][]int, n)
	for i := range b {
		b[i] = []int{i}
	}
	return b
}

func (mat *Matrix) find(n int) int {
	for i := range mat.blocks {
		for j := range mat.blocks[i] {
			if mat.blocks[i][j] == n {
				return i
			}
		}
	}
	panic("not found")
}

func (mat *Matrix) reset(n int) {
	stack := mat.blocks[mat.find(n)]

	for last := len(stack) - 1; stack[last] != n; last = len(stack) - 1 {
		i := stack[last]
		mat.blocks[i] = append(mat.blocks[i], i)
		stack = stack[:last]
	}
}

func (mat *Matrix) move(a, b int) {
	p1, p2 := mat.find(a), mat.find(b)
	mat.blocks[p1] = mat.blocks[p1][:len(mat.blocks[p1])-1]
	mat.blocks[p2] = append(mat.blocks[p2], a)
}

func (mat *Matrix) pile(a, b int) {
	p1, p2 := mat.find(a), mat.find(b)
	for i, val := range mat.blocks[p1] {
		if val == a {
			blocksToMove := mat.blocks[p1][i:]
			mat.blocks[p1] = mat.blocks[p1][:i]
			mat.blocks[p2] = append(mat.blocks[p2], blocksToMove...)
			return
		}
	}
}

func (mat *Matrix) process(act, pos string, a, b int) {
	if act == "move" {
		mat.reset(a)
	}
	if pos == "onto" {
		mat.reset(b)
	}

	switch act {
	case "move":
		mat.move(a, b)
	case "pile":
		mat.pile(a, b)
	}
}

func run() {
	f, _ := os.Open("input.txt")
	defer f.Close()

	var n, a, b int
	var act, pos string
	fmt.Fscan(f, &n)

	mat := &Matrix{blocks: newBlocks(n)}

	for {
		if fmt.Fscan(f, &act); act == "quit" {
			break
		}
		fmt.Fscan(f, &a, &pos, &b)
		mat.process(act, pos, a, b)
	}

	for i, line := range mat.blocks {
		fmt.Printf("%d:", i)
		for _, val := range line {
			fmt.Printf(" %d", val)
		}
		fmt.Println()
	}
}

func main() {
	run()
}

File: UVa 101 The Blocks Problem/main.py


class Blocks:

    def __init__(self, n):
        self.blocks = [[i] for i in range(n)]

    def find(self, val):
        """ Return line index that has value. """
        for i, line in enumerate(self.blocks):
            if val in line:
                return i
        assert False, "not found"
        return -1

    def reset(self, val):
        pos = self.find(val)
        stack = self.blocks[pos]

        while stack[-1] != val:
            self.blocks[stack[-1]].append(self.blocks[pos][-1])
            stack = stack[:-1]

    def process(self, action, src, directive, dest):
        if action == "move":
            self.reset(src)

        if directive == "onto":
            self.reset(dest)

        if action == "move":
            from_, to = self.find(src), self.find(dest)
            self.blocks[from_] = self.blocks[from_][:-1]
            self.blocks[to].append(src)

        elif action == "pile":
            from_, to = self.find(src), self.find(dest)
            for i, val in enumerate(self.blocks[from_]):
                if val == src:
                    blocksToMove = self.blocks[from_][i:]
                    self.blocks[from_] = self.blocks[from_][:i]
                    self.blocks[to].extend(blocksToMove)
                    break

    def result(self):
        for i, line in enumerate(self.blocks):
            print(i, ":", " ".join(str(b) for b in line))


def main():
    with open("input.txt") as f:
        n = int(f.readline())
        blocks = Blocks(n)

        for line in f:
            if line == "quit":
                break

            action, from_, directive, to = (a for a in line.split())
            from_, to = int(from_), int(to)
            blocks.process(action, from_, directive, to)

        blocks.result()

if __name__ == "__main__":
    main()
